web_services/profile/app.js

```diff
@@ -33,149 +35,209 @@ app.post("/upload/:channel", function (req, res) {
-    if (current_data.length >= config.channel_entry_limit && !isRecoveryChannel(uuid)) {
-      res.status(config.channel_limit_response_code).json({error: config.channel_limit_message});
-      return;
+  let channel = channelCache.get(channelId)
+  if (!channel) {
+    // Create new channel.
+    channel = {
+      cache: new NodeCache(config.channel_entry_config),
+      size: 0,
+      ttl
     }
-
-    // add data to channel
-    cacheEntry = current_data.concat([profile]);
+    // save channel in cache with requested TTL
+    channelCache.set(channelId, channel, ttl)
   } else {
-    // this is an initial profile upload. Create new channel.
-    cacheEntry = [profile]
+    // existing channel. check if this channel was about to expire, but got another upload
+    if (channel.cache.keys().length === 0) {
+      console.log(`Restoring requested TTL ${channel.ttl} for channel ${channelId}`)
+      channelCache.ttl(channelId, channel.ttl)
+    }
   }
 
-  // save data in cache
-  dataCache.set(channel, cacheEntry, async function (err, success) {
-    if (err) {
-      console.log(err);
-      res.status(500).json({ error: "unable to store profile data" });
-      return;
+  // Check if there is already data with the provided uuid to prevent duplicates
+  const existingData = channel.cache.get(uuid)
+  if (existingData) {
+    if (existingData === data) {
+      console.log(`Received duplicate profile ${uuid} for channel ${channelId}`)
+      // restart TTL counter of channel
+      // TODO: Not needed anymore? channel has default lifetime of 24 hours!
+      channelCache.ttl(channelId, channel.requestedTtl)
+      res.status(201).json({ success: true });
+    } else {
+      // Same UUID but different content? This is scary. Likely client bug. Bail out.
+      res.status(500).json({error: `Profile ${uuid} already exists in channel ${channelId} with different data.`});
     }
-    res.status(201);
-    res.json({ success: true });
-  });
-});
-
-app.post("/upload", function (req, res, next) {
-  const { data, uuid } = req.body;
+    return;
+  }
 
-  if (!uuid) {
-    res.status(404).json({ error: "missing uuid" });
+  // check channel size
+  const entrySize = sizeof(data) + sizeof(uuid)
+  const newSize = channel.size + entrySize
+  console.log(`channel ${channelId} newSize: ${newSize},\t delta: ${entrySize} bytes`)
+  if (newSize > config.channel_max_size_bytes) {
+    // channel full :-(
+    res.status(config.channel_limit_response_code).json({error: config.channel_limit_message});
     return;
   }
 
   // save data in cache
-  if (data) {
-    dataCache.set(uuid, data, async function (err, success) {
-      if (err) {
-        console.log(err);
-        res.status(500).json({ error: "unable to store profile data" });
-        return;
-      }
-      res.json({ success: true });
-    });
-  } else {
-    res.status(404).json({ error: "missing data" });
+  try {
+    channel.cache.set(uuid, data)
+    channel.size = newSize
```

Comment:
Is not it required to use `channelCache.set(channelId, channel)` to reset the value of the channel?
My question seems to be related to the above question about the possibility of recursive NodeCache objects and how it stores values.